src/groups/mqb/mqbblp/mqbblp_queueconsumptionmonitor.cpp

```diff
@@ -199,99 +200,372 @@ void QueueConsumptionMonitor::reset()
 
-    if (BSLS_PERFORMANCEHINT_PREDICT_UNLIKELY(d_maxIdleTime == 0)) {
+    if (d_alarmEventHandle || d_maxIdleTimeSec == 0) {
+        // Event is already scheduled or monitoring is disabled.
+        return;  // RETURN
+    }
+
+    // Schedule the event to be executed in 'now + maxIdleTime' time.
+    scheduleAlarmEvent(
+        calculateAlarmTime(0, d_queueState_p->scheduler()->now()));
+}
+
+void QueueConsumptionMonitor::onMessageSent(const bsl::string& appId)
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+
+    if (BSLS_PERFORMANCEHINT_PREDICT_UNLIKELY(d_maxIdleTimeSec == 0)) {
         // monitoring is disabled
         BSLS_PERFORMANCEHINT_UNLIKELY_HINT;
         return;  // RETURN
     }
 
-    BSLS_ASSERT_SAFE(currentTimer >= d_currentTimer);
+    SubStreamInfo& info = subStreamInfo(appId);
+    if (info.d_state == State::e_IDLE) {
+        // Call callback to check un-delivered messages with alarm time in the
+        // past.
+        bsls::TimeInterval now = d_queueState_p->scheduler()->now();
+        bsls::TimeInterval alarmTime;
+        bslma::ManagedPtr<mqbi::StorageIterator> oldestMsgIt =
+            d_haveUndeliveredCb(&alarmTime, appId, now);
+        if (oldestMsgIt) {
+            if (alarmTime <= now) {
+                // There are un-delivered messages with alarm time in the past.
+                // Since the idle event was scheduled, it will continue to
+                // track state.
+                BSLS_ASSERT_SAFE(info.d_idleEventHandle);
+
+                return;  // RETURN
+            }
+
+            // Since there is the oldest un-delivered message with alarm time
+            // in the future, schedule or reschedule the alarm event if needed.
+            scheduleOrRescheduleAlarmEventIfNeeded(alarmTime);
+        }
+
+        // There are no un-delivered messages or alarm time of the oldest
+        // message is in the future, so we can transition the substream to
+        // alive state.
+        onTransitionToAlive(&info, appId);
+    }
+}
+
+void QueueConsumptionMonitor::onTransitionToAlive(SubStreamInfo* subStreamInfo,
+                                                  const bsl::string& appId)
+{
+    // executed by the *QUEUE DISPATCHER* thread
+
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(d_queueState_p->queue()->dispatcher()->inDispatcherThread(
+        d_queueState_p->queue()));
+    BSLS_ASSERT_SAFE(subStreamInfo->d_state == State::e_IDLE);
+
+    if (subStreamInfo->d_idleEventHandle) {
+        // Cancel the idle event if it was scheduled.
+        d_queueState_p->scheduler()->cancelEventAndWait(
+            &subStreamInfo->d_idleEventHandle);
+        subStreamInfo->d_idleEventHandle.release();
+    }
+
+    subStreamInfo->d_state = State::e_ALIVE;
+
+    bdlma::LocalSequentialAllocator<2048> localAllocator(0);
```

Comment:
1. Let's pass the allocator passed on `QueueConsumptionMonitor` construction?
2. I think it's enough to have a block of <1024> bytes for both `UriBuilder` and `Uri`. Maybe even 512 is enough, but I am less sure.